packages/react-widgets/src/models/TimeSeriesModel.js

```diff
@@ -1,22 +1,212 @@
+    filtersLogicalOperator: source.filtersLogicalOperator,
     column,
     stepSize,
     operationColumn: operationColumn || column,
     joinOperation,
     operation
   });
+}
+
+// From remote
+const STEP_SIZE_TO_GROUP_QUERY_MAP = {
+  [GroupDateTypes.YEARS]: [GroupDateTypes.YEARS],
+  [GroupDateTypes.MONTHS]: [GroupDateTypes.YEARS, GroupDateTypes.MONTHS],
+  [GroupDateTypes.DAYS]: [
+    GroupDateTypes.YEARS,
+    GroupDateTypes.MONTHS,
+    GroupDateTypes.DAYS
+  ],
+  [GroupDateTypes.HOURS]: [
+    GroupDateTypes.YEARS,
+    GroupDateTypes.MONTHS,
+    GroupDateTypes.DAYS,
+    GroupDateTypes.HOURS
+  ],
+  [GroupDateTypes.MINUTES]: [
+    GroupDateTypes.YEARS,
+    GroupDateTypes.MONTHS,
+    GroupDateTypes.DAYS,
+    GroupDateTypes.HOURS,
+    GroupDateTypes.MINUTES
+  ]
+};
+
+const FORMAT_NAME_BY_STEP_SIZE = {
+  [GroupDateTypes.YEARS]: (row) =>
+    Date.UTC(row[formatStepSizeColumn(GroupDateTypes.YEARS)], 0),
+  [GroupDateTypes.MONTHS]: (row) =>
+    Date.UTC(
+      row[formatStepSizeColumn(GroupDateTypes.YEARS)],
+      row[formatStepSizeColumn(GroupDateTypes.MONTHS)] - 1
+    ),
+  [GroupDateTypes.DAYS]: (row) =>
+    Date.UTC(
+      row[formatStepSizeColumn(GroupDateTypes.YEARS)],
+      row[formatStepSizeColumn(GroupDateTypes.MONTHS)] - 1,
+      row[formatStepSizeColumn(GroupDateTypes.DAYS)]
+    ),
+  [GroupDateTypes.HOURS]: (row) =>
+    Date.UTC(
+      row[formatStepSizeColumn(GroupDateTypes.YEARS)],
+      row[formatStepSizeColumn(GroupDateTypes.MONTHS)] - 1,
+      row[formatStepSizeColumn(GroupDateTypes.DAYS)],
+      row[formatStepSizeColumn(GroupDateTypes.HOURS)]
+    ),
+  [GroupDateTypes.MINUTES]: (row) =>
+    Date.UTC(
+      row[formatStepSizeColumn(GroupDateTypes.YEARS)],
+      row[formatStepSizeColumn(GroupDateTypes.MONTHS)] - 1,
+      row[formatStepSizeColumn(GroupDateTypes.DAYS)],
+      row[formatStepSizeColumn(GroupDateTypes.HOURS)],
+      row[formatStepSizeColumn(GroupDateTypes.MINUTES)]
+    )
 };
+
+function groupWeeklyAverageValuesByDateColumn(data) {
+  const groups = data.reduce((acc, item) => {
+    const value = item.name;
+    const formattedValue = new Date(value);
+    const groupKey = getMonday(formattedValue);
+
+    if (!isNaN(groupKey)) {
+      let groupedValues = acc.get(groupKey);
+      if (!groupedValues) {
+        groupedValues = [];
+        acc.set(groupKey, groupedValues);
+      }
+
+      const aggregatedValue = item.value * item._grouped_count;
+
+      const isValid = aggregatedValue !== null && aggregatedValue !== undefined;
+
+      if (isValid) {
+        groupedValues.push([aggregatedValue, item._grouped_count]);
+        acc.set(groupKey, groupedValues);
+      }
+    }
+
+    return acc;
+  }, new Map());
+
+  return [...groups.entries()].map(([name, value]) => ({
+    name,
+    value:
+      value.filter((_, idx) => idx % 2 === 0).reduce((acc, [val]) => acc + val, 0) /
+      value.filter((_, idx) => idx % 2 !== 0).reduce((acc, [val]) => acc + val, 0)
+  }));
```

Comment:
This final part is a bit obscure, can you comment (or use some better variable names)?

Reply:
I agree. I'm going to improve it now.